File: cpp/leetcodes/Sort_List.cpp
/*
 *Sort a linked list in O(n log n) time using constant space complexity.

*/
#include <iostream>
#include <vector>
#include <bits/stl_algo.h>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <regex>


using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (head == nullptr){
            return nullptr;
        }
        if (head->next == nullptr){
            return head;
        }
        ListNode* slow = head;
        ListNode* fast = head->next;
        while (slow && fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* h2 = slow->next;
        slow->next = nullptr;
        auto h1 = sortList(head);
        h2 = sortList(h2);
        return mergeList(h1, h2);
    }
    ListNode* mergeList(ListNode* h1, ListNode* h2) {
        if (h1 == nullptr){
            return h2;
        }else if (h2 == nullptr) {
            return h1;
        }
        ListNode* res = new ListNode(0);
        ListNode* node = res;
        while (h1 && h2){
            if (h1->val < h2->val){
                node->next = h1;
                h1 = h1->next;
            }else{
                node->next = h2;
                h2 = h2->next;
            }
            node = node->next;
        }
        if (h1 != nullptr){
            node->next = h1;
        }else{
            node->next = h2;
        }
        return res->next;
    }

    void outputList(ListNode* head){
        ListNode* node = head;
        while (node != nullptr){
            cout << node->val << ", ";
            node = node->next;
        }
        cout << endl;
    }
};

int main(){
    Solution s;
    ListNode* h1 = new ListNode(1);
    h1->next = new ListNode(3);
    h1->next->next = new ListNode(5);
    ListNode* h2 = new ListNode(2);
    h2->next = new ListNode(3);
    h2->next->next = new ListNode(4);
    s.outputList(s.mergeList(h1, h2));


    h1 = new ListNode(3);
    h1->next = new ListNode(2);
    h1->next->next = new ListNode(4);
    s.outputList(s.sortList(h1));

    h1 = new ListNode(2);
    s.outputList(s.sortList(h1));


}

File: leetcode/medium/Sort_List.py
# -*- coding: utf-8 -*-

"""
created by huash06 at 2015-04-27 20:47

Sort a linked list in O(n log n) time using constant space complexity.

"""

__author__ = 'huash06'

import sys
import os
import itertools
import collections
import functools
import bisect
import datetime


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


    def output(self):
        print(self.val, end='->')
        if self.next:
            self.next.output()

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def sortList(self, head):
        if not head:
            return None
        l = 0
        h = head
        while h:
            l += 1
            h = h.next
        return self.mergeSort(head, l)


    def mergeSort(self, head, lenlist):
        if lenlist == 1:
            return head
        elif lenlist == 2:
            if head.val > head.next.val:
                tmp = head.val
                head.val = head.next.val
                head.next.val = tmp
            return head

        h = head
        l = 1
        while l < lenlist // 2:
            h = h.next
            l += 1
        nh = h.next
        h.next = None
        h1 = self.mergeSort(head, l)
        h2 = self.mergeSort(nh, lenlist-l)
        return self.mergeList(h1, h2)


    def mergeList(self, head1, head2):
        if not head2:
            return head1
        elif not head1:
            return head2

        ret = ListNode(-1000000)
        h1 = head1
        h2 = head2
        h = ret
        while h1 and h2:
            if h1.val < h2.val:
                h.next = h1
                h1 = h1.next
            else:
                h.next = h2
                h2 = h2.next
            h = h.next
        if h1:
            h.next = h1
        if h2:
            h.next = h2

        return ret.next



s = Solution()
s.sortList(ListNode(0)).output()
print()

l = ListNode(1)
l.next = ListNode(4)
node = l.next
vals = [3, 2, 5, 2]
for val in vals:
    node.next = ListNode(val)
    node = node.next
l.output()
print()
s.sortList(l).output()
print()
